### PR TITLE
Feature/mocha test reporter

### PR DESCRIPTION
**Pull request by chrisdp**
image

If you would like to test the new reporter simply add the `reporters` array show below to your `rooibos` config object in your `bsconfig.json` file. Also if you want to enable asci colorization see the setting below as well:
```json
  "rooibos": {
    ...
    "colorizeOutput": true,
    "reporters": [
      "mocha"
    ]
  }
```

**instillation**
If you would like to install a version of this branch we have an automation that automatically creates builds and posts the installation instructions as a comment on this pull request.

**Changes:**
- [x] Git diff/patch generation logic
- [x] Format patch into readable subset diff of actual and expected
- [x] Support colorization
- [x] Setting to disable colorization
- [x] Add new short hand to reporters array for `mocha`
- [x] Create new `MochaTestReporter`
- [x] Add execution time to test cases based on minimum threshold
- [x] Update docs to show new `mocha` reporter option
- [x] Create new `onSuiteBegin`, `onTestGroupBegin`, `onTestBegin`, `onTestComplete`, `onTestGroupComplete`, `onSuiteComplete` reporter hooks
- [x] migrate existing status logs into the `ConsoleTestReporter` using new hooks
- [x] Correct `ConsoleTestReporter` and `JUnitTestReporter` to use new event interfaces where applicable
- [x] Move debug logs not part of the reporters to new error, warning, info, debug, and trace log functions that are controlled with constants in the `CommonUtils.bs`
- [x] Add support for actual and expected strings to `TestResults` and the fail functions
- [x] Create new `asMultilineString` function
- [x] Update Assert functions to populate the actual and expected failure value where applicable
- [x] Update assertion failure messages to be more consistent and clear
- [x] Write failed examples for all updated Assertion functions
- [x] Add more tests for better coverage of when assertions should pass
- [x] Added new rooibos logs that will always log the `PASS`/`FAIL` status and that rooibos has shutdown after all reporters have logged their results
- [x] Updated CLI to use this new message allowing you to use other reporters other than just the `ConsoleTestReporter` with the CLI
- [x] Ignored tests are now iterated over for reporters to be able to log them in their results
- [x] Crashes in Assertions will now always report as a crash failure rather than a normal failure
- [x] Document new `@slow()` annotation

**Fixes:**
- [x] uninitialized can now be tested with `eqValues`
- [x] Fixed many crashes in assertions that could lead to false failures
- [x] Fixed some assertions that appeared to be broke (mostly around node tests)
- [x] Fixed some debugging logs that would state an error had occurred but in reality the tests where running fine
- [x] Custom messages are always passed if supplied
- [x] `failCrash` will use the message on the exception object unless a custom is supplied rather then always just reporting as `"Error"`

### CODE REVIEW

File: framework/src/source/BaseTestReporter.bs [hunk 1]
```diff
@@ -17,10 +17,36 @@ namespace rooibos
 
     function onBegin(ev as dynamic)
       'override me
+      ? "BaseTestReporter.onBegin"
+    end function
+
+    function onSuiteBegin(suite as rooibos.BaseTestSuite)
+      'override me
+    end function
```
Comment (chrisdp) [inline framework/src/source/BaseTestReporter.bs]:
```brighterscript
    interface TestReporterOnSuiteBeginEvent
      suite as rooibos.BaseTestSuite
    end interface
'.....
    function onSuiteBegin(event as rooibos.TestReporterOnSuiteBeginEvent)
      'override me
    end function
```
expand adding interfaces and event objects to the other hook

File: framework/src/source/CommonUtils.bs [hunk 2]
```diff
@@ -632,6 +703,31 @@ namespace rooibos.common
       return t
     end if
   end function
+
+  function getTypeWithComponentWrapper(value, includeSubtype = false) as string
```
Comment (chrisdp) [inline framework/src/source/CommonUtils.bs]: add doc block

File: framework/src/source/CommonUtils.bs [hunk 3]
```diff
@@ -632,6 +703,31 @@ namespace rooibos.common
...
+    else
+      if includeSubtype and rooibos.common.isSGNode(value)
+        return `<Component: ${type(value)}:${value.subType()}>`
+      else
+        return `<Component: ${type(value)}>`
+      end if
+    end if
+  end function
+
+  function truncateString(value as string, length = 38 as integer, collapseNewlines = true as boolean) as string
```
Comment (chrisdp) [inline framework/src/source/CommonUtils.bs]: add doc block

File: framework/src/source/MochaTestReporter.bs [hunk 4]
```diff
@@ -0,0 +1,845 @@
...
+            end if
+
+            resultMessage += chr(10)
+
+            if test.isParamTest
+              resultMessage += `${string(1, chr(9))}params at (file://${test.testSuite.filePath.trim()}:${Rooibos.Common.AsString(test.paramLineNumber + 1)})\n`
+              resultMessage += `${string(1, chr(9))}assertion at (file://${test.testSuite.filePath.trim()}:${Rooibos.Common.AsString(test.lineNumber)})\n`
+            else
+              resultMessage += `${string(1, chr(9))}at (file://${test.testSuite.filePath.trim()}:${Rooibos.Common.AsString(test.lineNumber)})\n`
+            end if
```
Comment (chrisdp) [inline framework/src/source/MochaTestReporter.bs]: think about adding this on crash cases as well.

File: framework/src/source/TestGroup.bs [hunk 5]
```diff
@@ -220,6 +228,37 @@ namespace rooibos
       return false
     end function
 
+    private sub notifyReportersOnTestGroupBegin()
```
Comment (chrisdp) [inline framework/src/source/TestGroup.bs]: possibly condense based on feedback.

File: framework/src/source/BaseTestSuite.bs [hunk 6]
```diff
@@ -614,35 +608,28 @@ namespace rooibos
...
-    '  * @description Fail if the array has the key.
-    '  * @param {Dynamic} array - target array
+    '  * @description Fail if the aa has the key.
+    '  * @param {Dynamic} aa - target aa
     '  * @param {Dynamic} key - key name
     '  * @param {Dynamic} [msg=""] - alternate error message
     '  * @returns {boolean} - true if the assert was satisfied, false otherwise
     '  */
-    function assertAANotHasKey(array, key, msg = "") as dynamic
+    function assertAANotHasKey(aa, key, msg = "") as dynamic
```
Comment (chrisdp) [inline framework/src/source/BaseTestSuite.bs]: add tests for this one

File: framework/src/source/BaseTestSuite.bs [hunk 7]
```diff
@@ -576,34 +575,29 @@ namespace rooibos
...
-    '  * @description Fail if the array doesn't have the key.
-    '  * @param {Dynamic} array - target array
+    '  * @description Fail if the aa doesn't have the key.
+    '  * @param {Dynamic} aa - target aa
     '  * @param {Dynamic} key - key name
     '  * @param {Dynamic} [msg=""] - alternate error message
     '  * @returns {boolean} - true if the assert was satisfied, false otherwise
     '  */
-    function assertAAHasKey(array, key, msg = "") as dynamic
+    function assertAAHasKey(aa, key, msg = "") as dynamic
```
Comment (chrisdp) [inline framework/src/source/BaseTestSuite.bs]: add tests for this one

File: framework/src/source/MochaTestReporter.bs (hunk 4 above)
Comment (chrisdp) [reply framework/src/source/MochaTestReporter.bs]: added

File: framework/src/source/TestGroup.bs (hunk 5 above)
Comment (chrisdp) [reply framework/src/source/TestGroup.bs]: Talked to @TwitchBronBron , We will do this later when we consolidating all the code flows as part of supporting promises.

File: framework/src/source/BaseTestSuite.bs [hunk 8]
```diff
@@ -1136,42 +1337,57 @@ namespace rooibos
...
+                end if
+              end if
+              m.currentResult.fail(msg, m.currentAssertLineNumber)
+              return false
+            end if
+          end for
+        else
+          ' I think we can remove this check, as we are already checking for valid types?
+          ' Will revisit this later.
+          throw `could not find comparator for type ${rooibos.common.asMultilineString(typeStr, true)}`
```
Comment (chrisdp) [inline framework/src/source/BaseTestSuite.bs]: Check if we still need this.

File: framework/src/source/BaseTestSuite.bs [hunk 9]
```diff
@@ -1526,7 +1809,9 @@ namespace rooibos
       try
         if type(node) = "roSGNode"
           if not rooibos.common.nodeContains(node, value)
-            msg = "Node doesn't have the '" + rooibos.common.asString(value, true) + "' value."
+            if msg = ""
+              msg = `expected "${rooibos.common.truncateString(rooibos.common.getTypeWithComponentWrapper(node, true))}" to contain child "${rooibos.common.truncateString(rooibos.common.getTypeWithComponentWrapper(value, true))}" by reference`
+            end if
             m.currentResult.fail(msg, m.currentAssertLineNumber)
```
Comment (chrisdp) [inline framework/src/source/BaseTestSuite.bs]: Triple check the intention of this assertion. It seems to look for a given child but the doc block makes it sound like something else.

File: framework/src/source/CommonUtils.bs [hunk 10]
```diff
@@ -863,5 +998,34 @@ namespace rooibos.common
...
+
+  sub logInfo(value)
+    #if ROOIBOS_INFO_LOGS
+      ? "[Rooibos Info]: " value
+    #end if
+  end sub
+
+  sub logDebug(value)
+    #if ROOIBOS_DEBUG_LOGS
+      ? "[Rooibos Debug]: "value
```
Comment (chrisdp) [inline framework/src/source/CommonUtils.bs]:
```suggestion
      ? "[Rooibos Debug]: " value
```

File: framework/src/source/CommonUtils.bs [hunk 11]
```diff
@@ -863,5 +998,34 @@ namespace rooibos.common
...
+    #end if
+  end sub
+
+  sub logTrace(value)
+    #if ROOIBOS_TRACE_LOGS
+      ? "[Rooibos Trace]: "value
```
Comment (chrisdp) [inline framework/src/source/CommonUtils.bs]:
```suggestion
      ? "[Rooibos Trace]: " value
```

File: framework/src/source/TestGroup.bs [hunk 12]
```diff
@@ -115,29 +129,38 @@ namespace rooibos
...
-      ' ? "Getting next async test"
+      rooibos.common.logTrace("Getting next async test")
       m.currentTestIndex++
       m.currentTest = m.tests[m.currentTestIndex]
       m.testSuite.isDoneCalled = false
       m.testTimer.mark()
       if m.currentTest = invalid
-        ' ? " all tests are finished"
+        rooibos.common.logTrace("All tests are finished")
         'finished
```
Comment (chrisdp) [inline framework/src/source/TestGroup.bs]:
```suggestion
```

File: bsc-plugin/src/lib/rooibos/CodeCoverageProcessor.spec.ts [hunk 13]
```diff
@@ -333,7 +333,7 @@ describe('RooibosPlugin', () => {
                 expect(a).to.equal(b);
             });
 
-            it('correctly transpiles some statements', async () => {
+            it.only('correctly transpiles some statements', async () => {
```
Comment (chrisdp) [inline bsc-plugin/src/lib/rooibos/CodeCoverageProcessor.spec.ts]:
```suggestion
            it.skip('correctly transpiles some statements', async () => {
```

File: bsc-plugin/src/lib/rooibos/CodeCoverageProcessor.spec.ts [hunk 14]
```diff
@@ -333,7 +333,7 @@ describe('RooibosPlugin', () => {
                 expect(a).to.equal(b);
             });
 
-            it('correctly transpiles some statements', async () => {
+            it.skip('correctly transpiles some statements', async () => {
```
Comment (chrisdp) [inline bsc-plugin/src/lib/rooibos/CodeCoverageProcessor.spec.ts]: Investigating a fix for this failing test with @TwitchBronBron

File: framework/src/source/BaseTestSuite.bs (hunk 8 above)
Comment (chrisdp) [reply framework/src/source/BaseTestSuite.bs]: WE can leave this is as a safely. I don't think we can hit it but it's also not hurting anything.

File: bsc-plugin/src/lib/rooibos/CodeCoverageProcessor.spec.ts (hunk 14 above)
Comment (chrisdp) [reply bsc-plugin/src/lib/rooibos/CodeCoverageProcessor.spec.ts]:
```suggestion
            it('correctly transpiles some statements', async () => {
```

Rolled back to brighterscript `0.67.4` till we can solve this issue caused by changes in `0.68`

File: framework/src/source/rooibos/BaseTestSuite.bs [hunk 15]
```diff
@@ -525,18 +541,18 @@ namespace rooibos
       end if
 
       try
-        if value <> invalid
+        if type(value) <> "Invalid"
```
Comment (luis-j-soares) [inline framework/src/source/rooibos/BaseTestSuite.bs]: This causes some failures.

```
    AssertionError: expected "roInvalid" to be invalid
```

It should be this instead:

```suggestion
        if rooibos.common.getSafeType(value) <> "Invalid"
```

**Review by luis-j-soares — CHANGES_REQUESTED** [review]

File: framework/src/source/rooibos/BaseTestSuite.bs (hunk 15 above)
Comment (chrisdp) [reply framework/src/source/rooibos/BaseTestSuite.bs]: - added new test that better replicates the `roInvalid` object when using `assertInvalid`
- fixed `assertInvalid` not passing `roInvalid`
- added new test for `assertNotInvalid`
- hardered `assertNotInvalid` to check the type in the same fasion as `assertInvalid` so it can't crash on `uninitialized`
